Zend/zend_ast.c

```diff
@@ -739,21 +740,56 @@ ZEND_API zend_result ZEND_FASTCALL zend_ast_evaluate(zval *result, zend_ast *ast
 				zend_error_noreturn(E_COMPILE_ERROR, "Cannot use [] for reading");
 			}
 
-			if (UNEXPECTED(zend_ast_evaluate(&op1, ast->child[0], scope) != SUCCESS)) {
+			zend_result evaluate_result;
+
+			evaluate_result = zend_ast_evaluate(&op1, ast->child[0], scope);
+			if (UNEXPECTED(evaluate_result) != SUCCESS) {
 				ret = FAILURE;
-			} else if (UNEXPECTED(zend_ast_evaluate(&op2, ast->child[1], scope) != SUCCESS)) {
+				break;
+			}
+
+			// DIM on enums is disallowed because it allows executing arbitrary expressions
+			if (Z_TYPE(op1) == IS_OBJECT) {
+				zend_class_entry *ce = Z_OBJCE(op1);
+				if (ce->ce_flags & ZEND_ACC_ENUM) {
+					zend_error_noreturn(E_COMPILE_ERROR, "Cannot use [] on enums in constant expression");
```

Comment:
So it seems like it would be safe to throw an Error (instead of a fatal error) for opcache, etc. zend_ast_evaluate (and callers such as `zval_update_constant*`) already have to throw for using invalid array offsets, undeclared constants, etc., and will clear the exception if opcache is trying to evaluate a constant early as an optimization. (e.g. when opcache is compiling `$var = self::FOO[0];`)

But for a zend_error_noreturn, future/current opcache optimizations may cause snippets such as `$var = self::FOO[0]` to be a fatal error before that line is executed